Registry/RepositoryRegistry.php

```diff
@@ -0,0 +1,141 @@
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get($instanceName)
+    {
+        if (!isset($this->instances[$instanceName])) {
+            throw new \InvalidArgumentException(sprintf(
+                'Repository instance "%s" has not been registered, available repository instances: "%s"',
+                $instanceName,
+                implode('", "', array_keys($this->instances))
+            ));
+        }
+
+        return $this->instances[$instanceName];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepositoryAlias(ResourceRepository $repository)
+    {
+        foreach ($this->instances as $alias => $repositoryInstance) {
+            if ($repositoryInstance === $repository) {
+                return $alias;
+            }
+        }
+
+        throw new \InvalidArgumentException(sprintf(
+            'Unknown repository instance of type "%s", cannot determine the alias.',
+            get_class($repository)
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepositoryType(ResourceRepository $repository)
+    {
+        $repositoryClass = get_class($repository);
+        if (!isset($this->typeMap[$repositoryClass])) {
+            throw new \InvalidArgumentException(sprintf(
+                'No repository has been registered of class "%s", cannot determine the type, known types: "%s".',
+                $repositoryClass,
+                implode('", "', $this->typeMap)
+            ));
+        }
+
+        return $this->typeMap[$repositoryClass];
+    }
+
+    private function createRepositoryInstance($instanceName, array $config)
+    {
+        if (!isset($config['type'])) {
+            throw new \InvalidArgumentException(sprintf(
+                'Each instance configuration must have a "type" key (for configuration "%s")',
+                $instanceName
+            ));
+        }
+
+        $type = $config['type'];
+        $config = $config['options'];
+
+        if (!isset($this->factories[$type])) {
+            throw new \InvalidArgumentException(sprintf(
+                'Unknown repository type "%s", known types: "%s"',
+                $type, implode('", "', array_keys($this->factories))
+            ));
+        }
+
+        $factory = $this->factories[$type];
+        $defaultConfig = $factory->getDefaultConfig();
+
+        $configDiff = array_diff(
+            array_keys($config),
+            array_keys($defaultConfig)
+        );
+
+        if ($configDiff) {
+            throw new \InvalidArgumentException(sprintf(
+                'Invalid configuration keys "%s" for repository type "%s", valid config keys: "%s"',
+                implode('", "', $configDiff),
+                $type,
+                implode('", "', array_keys($defaultConfig))
+            ));
+        }
+
+        $config = array_merge($defaultConfig, $config);
+
+        if ($factory instanceof CompositeFactory) {
+            $factory->setRepositoryRegistry($this);
```

Comment:
The `CompositeFactory` requires an instance of the `RepositoryRegistry`, which requires an instance of the `CompositeFactory` ... so constructor injection of the registry to the factory will not work.

There are then 2 solutions that I saw: 1. Passing the registry as a second argument to `create` and 2. Doing this.

Rather than create a `RepositoryRegistryAwareInterface` I just type check on the `CompositeFactory` - it is a special case at the moment, if needs require we can add an `aware` interface later.